game-core/src/test/java/games/strategy/triplea/delegate/battle/MustFightBattleExecutablesTest.java

```diff
@@ -0,0 +1,378 @@
+
+    assertThat(
+        "FireNavalBombardment should be added for first round",
+        getIndex(execs, MustFightBattle.FireNavalBombardment.class), is(0));
+  }
+
+  @Test
+  @DisplayName("Verify Bombard step is not added on subsequent rounds")
+  void bombardStepNotAddedOnSubsequentRound() {
+    final MustFightBattle battle = newBattle(LAND);
+
+    battle.setUnits(
+        List.of(),
+        List.of(),
+        List.of(),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(false);
+
+    assertStepIsMissing(execs, MustFightBattle.FireNavalBombardment.class);
+  }
+
+  @Test
+  @DisplayName("Verify Bombard step is added but no bombardment happens if bombard units are empty")
+  void bombardStepAddedButNoBombardUnits() {
+    final MustFightBattle battle = newBattle(LAND);
+
+    battle.setUnits(
+        List.of(),
+        List.of(),
+        List.of(),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(true);
+
+    final int index = getIndex(execs, MustFightBattle.FireNavalBombardment.class);
+    final IExecutable step = execs.get(index);
+
+    final IDelegateBridge delegateBridge = mock(IDelegateBridge.class);
+    doAnswer(
+        invocation -> {
+          final Change change = invocation.getArgument(0);
+          mockGameData.performChange(change);
+          return null;
+        })
+        .when(delegateBridge)
+        .addChange(any());
+
+    step.execute(null, delegateBridge);
+
+    verify(mockGameData, times(1)).performChange(ChangeFactory.EMPTY_CHANGE);
+    // TODO: Somehow get the stack and check that no new execs were added to it.
+  }
+
+  @Test
+  @DisplayName("Verify Bombard step is added and bombardment happens if bombard units exist")
+  void bombardStepAddedAndBombardHappens() {
+    final MustFightBattle battle = newBattle(LAND);
+
+    when(unit1.getType()).thenReturn(mockUnit1Type);
+    when(unit1.getMovementLeft()).thenReturn(BigDecimal.ZERO);
+    final MutableProperty<Boolean> alreadyMovedProperty = MutableProperty.ofReadOnly(() -> true);
+    doReturn(alreadyMovedProperty).when(unit1).getPropertyOrThrow(ALREADY_MOVED);
+    when(mockUnit1Type.getAttachment(anyString())).thenReturn(mockUnit1Attachment);
+
+    when(unit2.getType()).thenReturn(mockUnit2Type);
+    when(unit2.getOwner()).thenReturn(defender);
+    when(mockUnit2Type.getAttachment(anyString())).thenReturn(mockUnit2Attachment);
+
+    battle.setUnits(
+        List.of(unit2),
+        List.of(),
+        List.of(unit1),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(true);
+
+    final int index = getIndex(execs, MustFightBattle.FireNavalBombardment.class);
+    final IExecutable step = execs.get(index);
+
+    final IDelegateBridge delegateBridge = mock(IDelegateBridge.class);
```

Comment:
I tried using `MockDelegateBridge#newDelegateBridge` but the strict stubbing check in Mockito makes it not usable.  Mockito complains if a stub was added that wasn't used and `newDelegateBridge` adds a lot of stubs that these individual tests don't use.

I could make turn on the `lenient` mode but I worry about those repercussions.  Ideas on how to simplify the mock setups?

Reply:
I think in part the issue is with the code that is being tested. It is doing too much, nearly everything, so we are left stubbing too many items and testing via `verify` and asserting side effects. 

An approach to explore, extract functionality from a top level into strategy objects. The strategy objects are then ideally nearly functional and require very few mocks to tests. The high level then contains most of the control-flow logic, easily mocked, and you mainly just need to assert interactions between objects at that level.

For example, if another object, a strategy object, were to compute the bombard steps, then you would not need to test for it here as that computation would be mocked. You'd only need to test that the mock object is invoked in the right conditions.

Reply:
So would you say it is fine to go with the current approach (complex mock setups) to ensure everything is tested?  And then as the code is reconstructed with helper functions, classes, etc, the mocks can get simplified.  I'd be fine with doing that.

I don't want to try and extract the functionality at this point because it is scattered all over the place.

Reply:
It's a chicken-and-egg problem. The code was not written with tests in mind, it is hardly testable, to make it testable, it needs to be re-written, to re-write it - it needs tests first.

A technique for working with this kind of code is to apply high level characterization and integration tests to test large modules. Re-write the inner ones carefully piece by piece and add unit tests to them. Combine with some manual testing and rinse-wash-repeat until it's all well decomposed. That can be a difficult process as the original code can have cyclical dependencies, poor modularity (or no modularity at all), etc..